centrallix/osdrivers/objdrv_mysql.c

```diff
@@ -1832,6 +1849,49 @@ mysd_internal_TreeToClause(pExpression tree, pMysdTable *tdata, pXString where_c
                     mysd_internal_TreeToClause((pExpression)(tree->Children.Items[0]), tdata,  where_clause,conn);
 		    xsConcatenate(where_clause, ") ", 2);
 		    }
+		else if (!strcmp(tree->Name,"upper"))
+		    {
+		    /** check if database is using an expected character set **/
+		    char* collation = NULL; 
+		    pMysdTable tempTdata = *tdata;
+
+		    if(strstr(tempTdata->Node->DatabaseCharset, "latin1") != NULL) collation = "latin1_bin";
+		    else if (strstr(tempTdata->Node->DatabaseCharset, "utf8mb4") != NULL) collation = "utf8mb4_bin";

```

Comment:
Do we also need to handle the utf8mb3 case here?  Any other cases?

(same for lower() below)

The upper/lower collation setting could also be detected and set during charset detection in GetConn(), so we don't need to have the logic twice here in the function special cases.

Reply:
I am unsure what collations would be useful/necessary, but most of the character sets listed as supported by MariaDB have a collation that is of the form <CHARSET NAME>_bin, by appending _bin to the end of the character set name we support almost everything. The only exceptions are the binary charset, which I added a check for, and the uca1400 collations, which do not appear to be supported by the version of MariaDB we use and have a character set of NULL; it is treated like an empty/failed request.

Good idea; I move the logic and store the collation rather than the character set now.